TrackInfo and derivative should be a class, not a struct.

jyavenard opened the pull request:
#### 740232e86f953e5363c619c318901fe848125733
<pre>
TrackInfo and derivative should be a class, not a struct.
<a href="https://bugs.webkit.org/show_bug.cgi?id=304329">https://bugs.webkit.org/show_bug.cgi?id=304329</a>
<a href="https://rdar.apple.com/166697941">rdar://166697941</a>

Reviewed by Youenn Fablet.

We make TrackInfo a class with immutable members.

Covered by existing tests.
* Source/WebCore/platform/MediaSamplesBlock.h:
(WebCore::MediaSamplesBlock::trackID const):
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::operator== const):
(WebCore::TrackInfo::codecName const):
(WebCore::TrackInfo::codecString const):
(WebCore::TrackInfo::trackID const):
(WebCore::TrackInfo::setTrackID):
(WebCore::TrackInfo::encryptionData const):
(WebCore::TrackInfo::encryptionOriginalFormat const):
(WebCore::TrackInfo::encryptionInitDatas const):
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::size const):
(WebCore::VideoInfo::displaySize const):
(WebCore::VideoInfo::bitDepth const):
(WebCore::VideoInfo::colorSpace const):
(WebCore::VideoInfo::extensionAtoms const):
(WebCore::VideoInfo::immersiveVideoMetadata const):
(WebCore::VideoInfo::toVideoInfoData const):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::create):
(WebCore::AudioInfo::rate const):
(WebCore::AudioInfo::channels const):
(WebCore::AudioInfo::framesPerPacket const):
(WebCore::AudioInfo::bitDepth const):
(WebCore::AudioInfo::cookieData const):
(WebCore::AudioInfo::toAudioInfoData const):
(WebCore::AudioInfo::AudioInfo):
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WebCore::createVideoInfoFromAV1CodecConfigurationRecord):
(WebCore::createVideoInfoFromAV1Stream):
* Source/WebCore/platform/graphics/AV1Utilities.h:
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
* Source/WebCore/platform/graphics/MediaSampleConverter.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:
(WebCore::AudioTrackPrivateWebM::codec const):
(WebCore::AudioTrackPrivateWebM::sampleRate const):
(WebCore::AudioTrackPrivateWebM::numberOfChannels const):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::audioStreamDescriptionFromAudioInfo):
(WebCore::getEncryptionData):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::PacketDurationParser::PacketDurationParser):
(WebCore::setEncryptionInfo): Deleted.
* Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromAVCC):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::returnSamples):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::setFormatDescription):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::codec const):
(WebCore::VideoTrackPrivateWebM::width const):
(WebCore::VideoTrackPrivateWebM::height const):
(WebCore::VideoTrackPrivateWebM::colorSpace const):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::createAudioInfoForFormat):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/304673@main">https://commits.webkit.org/304673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/319984448679efc8111ac6ea8a526a3bf3bd371c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143947 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138109 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8438 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104170 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139183 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122080 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85001 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/80f6c401-0d07-4641-9172-5be56bb631e1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6398 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4059 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4541 "Built successfully") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128195 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115685 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146693 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8277 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40848 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112511 "Found 7 new test failures: imported/w3c/web-platform-tests/css/css-text/hyphens/hyphenate-character-002.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-auto-010.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-auto-inline-010.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-auto-last-word-001.html imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-punctuation-001.html imported/w3c/web-platform-tests/css/filter-effects/filters-drop-shadow-002.html imported/w3c/web-platform-tests/quirks/line-height-in-list-item.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8293 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6952 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112852 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28643 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6329 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118385 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8325 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36442 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8043 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71884 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8264 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8117 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -103,27 +117,23 @@ struct TrackInfo : public ThreadSafeRefCounted<TrackInfo> {
-    }
+
 #if ENABLE(ENCRYPTED_MEDIA)
     TrackInfo(TrackType type, FourCC codecName, const String& codecString, TrackID trackID, std::optional<EncryptionData>&& encryptionData, std::optional<FourCC> encryptionOriginalFormat, Vector<EncryptionInitData>&& encryptionInitDatas)
```
I would be tempted to have a TrackInfoData with all these members, including the `#if ENABLE(ENCRYPTED_MEDIA)` ones.
Then we would just write:
```
TrackInfo(TrackType type, TrackInfoData&& data)
    : m_type(type)
    , m_data(WTFMove(data))
{
}
```

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -133,103 +143,125 @@ struct TrackInfo : public ThreadSafeRefCounted<TrackInfo> {
+};
 
-    FloatSize size;
+struct VideoInfoData {
```
I would tend to have VideoInfoData = std::pair<TrackInfoData, VideoSpecificInfoData> or a simple structure if is easier;
Then VideoInfo would have its `VideoSpecificInfoData m_data` member and you would not need to have these long constructors and long `to` methods.

jyavenard replied:
it would make the construction a tad more verbose with extra level of `{ }` ; I could just VideoInfo be a single member VideoInfoData ...

jyavenard committed:
TrackInfo and derivative should be a class, not a struct.
https://bugs.webkit.org/show_bug.cgi?id=304329
rdar://166697941

Reviewed by Youenn Fablet.

We make TrackInfo a class with immutable members.

Covered by existing tests.
* Source/WebCore/platform/MediaSamplesBlock.h:
(WebCore::MediaSamplesBlock::trackID const):
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::operator== const):
(WebCore::TrackInfo::codecName const):
(WebCore::TrackInfo::codecString const):
(WebCore::TrackInfo::trackID const):
(WebCore::TrackInfo::setTrackID):
(WebCore::TrackInfo::encryptionData const):
(WebCore::TrackInfo::encryptionOriginalFormat const):
(WebCore::TrackInfo::encryptionInitDatas const):
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::size const):
(WebCore::VideoInfo::displaySize const):
(WebCore::VideoInfo::bitDepth const):
(WebCore::VideoInfo::colorSpace const):
(WebCore::VideoInfo::extensionAtoms const):
(WebCore::VideoInfo::immersiveVideoMetadata const):
(WebCore::VideoInfo::toVideoInfoData const):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::create):
(WebCore::AudioInfo::rate const):
(WebCore::AudioInfo::channels const):
(WebCore::AudioInfo::framesPerPacket const):
(WebCore::AudioInfo::bitDepth const):
(WebCore::AudioInfo::cookieData const):
(WebCore::AudioInfo::toAudioInfoData const):
(WebCore::AudioInfo::AudioInfo):
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WebCore::createVideoInfoFromAV1CodecConfigurationRecord):
(WebCore::createVideoInfoFromAV1Stream):
* Source/WebCore/platform/graphics/AV1Utilities.h:
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
* Source/WebCore/platform/graphics/MediaSampleConverter.h:
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.cpp:
(WebCore::AudioTrackPrivateWebM::codec const):
(WebCore::AudioTrackPrivateWebM::sampleRate const):
(WebCore::AudioTrackPrivateWebM::numberOfChannels const):
* Source/WebCore/platform/graphics/cocoa/AudioTrackPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::audioStreamDescriptionFromAudioInfo):
(WebCore::getEncryptionData):
(WebCore::createAudioFormatDescription):
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createAudioInfoFromFormatDescription):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
(WebCore::PacketDurationParser::PacketDurationParser):
(WebCore::setEncryptionInfo): Deleted.
* Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromAVCC):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParser.h:
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.cpp:
(WebCore::WebMParser::VideoTrackData::consumeFrameData):
(WebCore::WebMParser::AudioTrackData::consumeFrameData):
(WebCore::SourceBufferParserWebM::returnSamples):
* Source/WebCore/platform/graphics/cocoa/SourceBufferParserWebM.h:
(WebCore::WebMParser::TrackData::setFormatDescription):
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::createVideoInfoFromVPCodecConfigurationRecord):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::codec const):
(WebCore::VideoTrackPrivateWebM::width const):
(WebCore::VideoTrackPrivateWebM::height const):
(WebCore::VideoTrackPrivateWebM::colorSpace const):
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/WebMAudioUtilitiesCocoa.mm:
(WebCore::createAudioInfoForFormat):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::processVideoEncoderActiveConfiguration):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addAudioTrack):
(WebCore::MediaRecorderPrivateWriterWebMDelegate::addVideoTrack):
* Source/WebCore/platform/video-codecs/cocoa/RTCVideoDecoderVTBAV1.mm:
(computeAV1InputFormat):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/304673@main
